src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -638,13 +638,7 @@ void ObjectSynchronizer::exit(oop object, BasicLock* lock, JavaThread* current)
   // The ObjectMonitor* can't be async deflated until ownership is
   // dropped inside exit() and the ObjectMonitor* must be !is_busy().
   ObjectMonitor* monitor = inflate(current, object, inflate_cause_vm_internal);
-  if (LockingMode == LM_LIGHTWEIGHT && monitor->is_owner_anonymous()) {
-    // It must be owned by us. Pop lock object from lock stack.
-    LockStack& lock_stack = current->lock_stack();
-    oop popped = lock_stack.pop();
-    assert(popped == object, "must be owned by this thread");
-    monitor->set_owner_from_anonymous(current);
-  }
+  assert(!monitor->is_owner_anonymous(), "must not be");
```

Comment:
Agreed that the block from old L641 -> L647 is impossible after the call
to `inflate()` on L640. Thanks for adding that assert.

The other reason is if the owner came in here with a lightweight monitor,
then that would have been handled via L571 -> L592 and we never should
have reached this block.